docs/sdg/sdg-refactor.md

```diff
@@ -0,0 +1,136 @@
+- As with Option 1, this approach suffers from the fact that keeping interconnected components in the same repository provides less pressure to consistently document the API contracts between them.  In the case of Option 1, the interconnected components that would not have as much pressure to be documented would be preprocessing/postprocessing and core SDG.  In the case of Option 3, the interconnected components that would not have as much pressure to be documented would be the CLI and preprocessing/postprocessing.  However, in both cases, this con could be alleviated by just having the discipline to document the APIs well even without such pressure.
+- As with Option 2, this approach would not enable user flow 2.1.  Maybe that's fine since it is not on our requirements list.
+
+Conclusion:
+
+- This seems like a reasonable option.  The cons are mostly manageable.  However, overall the pros of Option 1 seem more compelling.
+
+### Option 4: Preprocessing and postprocessing go to different locations
+
+We could also mix and match any of the above options separately for preprocessing and postprocessing.  For example, preprocessing could move to the CLI repo and postprocessing could stay in the SDG repo.  Or preprocessing could move to a new repository and postprocessing could move to a different new repository or the _same_ new repository.  Enumerating all possible permutations of where each could go and enumerating pros and cons of each of them would make this document unbearably long.  If anyone wants to advocate for a small number of specific permutations, we will add them to this document.
+
+## Question 4: Should preprocessing, postprocessing, and core SDG be separate Python packages?
+
+If we choose Option 1 (leave preprocessing and postprocessing in SDG) then we still have the option to separate them into distinct Python packages.  That would get us some of the benefits of Option 2 (moving them to a different repository) while avoiding *some* of the costs of Option 2.  In particular, it would make the boundaries clearer and put more pressure on the developers of preprocessing, postprocessing, and core SDG to have well documented contracts for how each of these elements interact.  With that said, it would also bring in some additional complexity and increase the amount of work involved.
+
+## Decisions
+
+Since this is a draft, no decisions are made yet.  However, here are the current draft decisions:
+
+- We will support the following user flows: 1.1., 1.2., 1.3, 1.3.1, 1.3.2, 2.1, and 2.1 as documented in the Question 1 section above.
+- We will adopt the updates to the CLI that will be documented in Question 2 above.
+- We will leave the preprocessing in SDG as described in Question 3: Option 1.
```

Comment:
Given my comments above about the intention of each library/repository at their conception and the future plan for what we hope each part of this project to become, I believe moving the data ingestion functionality to the CLI makes the most sense. 

the CLI is the data, model, and taxonomy manager. Having it be the document manager that feeds into all of the various other parts of the workflow logically makes sense to me.

Reply:
Moving more application logic into the CLI seems to b directly at odds with the intended future of having REST APIs. If we move more things into the CLI, wouldn't we shortly move them back out in order to have them be able to live behind a REST API?

Reply:
I guess my expectation is that the REST API implementation would either become _part_ of the CLI repo or would be a shim that calls into the CLI repo.  Either way, calling it the "CLI repo" seems misleading.  I would be much more comfortable with this change if we were calling that repo ( https://github.com/instructlab/instructlab ) the InstructLab _core_ repo or something like that instead of just "CLI repo".

Reply:
I would not expect REST APIs to be implemented in something called a CLI.

Hold on - when people say "the CLI" are they referring to the `instructlab/instructlab` repo as a whole? If so, I did not understand that at all and explains a bunch of my confusion!

Reply:
Yes, that's my understanding.  For example, when I click on https://github.com/instructlab/instructlab and then look at the readme, the title is "InstructLab 🐶 (ilab)" but the first section heading is "[Welcome to the InstructLab CLI](https://github.com/instructlab/instructlab#welcome-to-the-instructlab-cli)".  To me that suggests that the entire repo is referred to as the CLI.  I still think the name is misleading, but rethinking the naming of that repo would be outside the scope of this dev-doc.  Also, I can see a case for why from a _user_ standpoint, that name makes sense: if you install this piece of software then you can run InstructLab via the CLI.  It just doesn't make a lot of sense from a developer standpoint, which is normally the perspective I would expect to drive decisions about github repository names.

Reply:
yeah @anastasds  I think this statement is a part of the confusion:
`I would not expect REST APIs to be implemented in something called a CLI.`

`instructlab/instructlab` is referred to as the CLI (though it is much much much more), since it was once called the `instructlab/cli` repo in olden times

The planning for the REST API is for it to "wrap" this repo, given the substantial work going on to re-design it in a way that works better with an api:
https://github.com/instructlab/instructlab/issues/1548
https://github.com/instructlab/instructlab/issues/2325

your comment about moving things out of "the cli". This will most likely, never happen. If that were to happen, we would need to have a different repo for each command and workflow currently living in the "cli".

I think this point right here is a large part of the contention. In my planning for the engine runtime team, adding functionality like this makes _sense_ given that `instructlab/instructlab` is the orchestrator of data, models, config, taxonomy, etc. Having it be the orchestrator of documents and even more taxonomy workflows fits directly into the design of that repo

Reply:
I thin planning where this functionality should go based on the idea of moving core functionality out of `instructlab/instructlab` is a misnomer as long term planning of the instruct lab (even with a rest api) does not necessitate moving anything out of the repo. An example I like to use often is: https://github.com/containers/podman

Podman seamlessly maintains a user facing CLI and a rest api in the same repo that allows them to streamline development  and support various use cases of folks accessing Podman via `podman-remote`

again though, ** if the only idea against this functionality living in the instruct lab repo is the REST API, I think this is both an unrelated topic and a misnomer, as the team aiming to implement the REST API: @jaideepr97 @alinaryan, me @nathan-weinberg, have made no plans to move any functionality out of the repo.

this would necessitate  things like:

`instructlab/download`
`instructlab/config`
`instructlab/chat`
`instructlab/serve`

which are repos that I think introduce an extreme level of modularization

Reply:
I posted a comment [in the CLI 1.0 dev doc](https://github.com/instructlab/dev-docs/pull/159#discussion_r1850587382) advocating for putting an end to using the phrase "the CLI" to refer to `instructlab/instructlab`.  I am hoping that if we can resolve that issue there, that this can be a reasonable conclusion to this thread too.  Let's see how that one turns out.